### DISCUSSION TITLE
How to trigger IntelliSense for a restart after cmake configuration from customized task?

### DISCUSSION THREAD

**Discussion by zchrissirhcz** [microsoft/vscode-cmake-tools#3875] [Q&A] [answered]
I'd like to passing args for "CMake: Configure" task, hence I write one in `tasks.json`.

The "CMake: Configure" template task also generates cmake-file-api files (`build/.cmake/api/v1/query/client-vscode/query.json`), and I figured it out they can be generated via command line (from https://github.com/Decodetalkers/neocmakelsp/issues/44).

So I create two tasks in `tasks.json`:  `Generate cmake-file-api files` and `cmake-configure`.

```json
{
	"version": "2.0.0",
	"tasks": [
        {
            "label": "Generate cmake-file-api files",
            "type": "shell",
            "command": "zsh",
            "args": [
                "-c",
                "mkdir -p build/.cmake/api/v1/query/client-vscode && cmake -E capabilities | jq '.fileApi' > build/.cmake/api/v1/query/client-vscode/query.json"
            ],
            "problemMatcher": []
        },
        {
            "label": "cmake-configure",
            "type": "shell",
            "command": "zsh",
            "args": [
                "-c",
                "cmake -S . -B build -DHELLO_TYPE=HAPPY",
            ],
            "dependsOn":[
                "Generate cmake-file-api files"
            ],
            "options": {
                "cwd": "${workspaceFolder}"
            },
            "problemMatcher": [],
        },
    ]
}
```

The task `cmake-configure` will trigger `Generate cmake-file-api files` and run without error, but there is still red squirrels in C++ files for 3rdparty header files, and Ctrl+Mouse Click cannot go to function declarations. I guess a restart of IntelliSense is required, but I don't know if it can be "restarted", just like the VSCode Clangd Extension's `Clangd: Restart Clangd Server`. If it is possible, then how can I write it in `tasks.json`?

**Comment by zchrissirhcz** [comment]
To reproduce, use these files:

`main.cpp`:
```cpp
#include <stdio.h>
#include <opencv2/opencv.hpp>

int main(int argc, char** argv)
{
    printf("Hello, from hello!\n");
    printf("args:");
    for (int i=0; i<argc; i++)
    {
        printf(" %s", argv[i]);
    }
    printf("\n");

    const std::string image_path = "/Users/zz/data/1920x1080.jpg";
    cv::Mat image = cv::imread(image_path);
    cv::imshow("image", image);
    cv::waitKey(0);

    return 0;
}
```

`CMakeLists.txt`:
```cmake
cmake_minimum_required(VERSION 3.5)
project(hello VERSION 0.1.0 LANGUAGES C CXX)

find_package(OpenCV REQUIRED)
add_executable(hello main.cpp)
target_link_libraries(hello PRIVATE ${OpenCV_LIBS})

message(STATUS "HELLO_TYPE: ${HELLO_TYPE}")
```

**Comment by zchrissirhcz** [comment]
The actual VSCode is:

![image](https://github.com/microsoft/vscode-cmake-tools/assets/3831847/28b7b7bc-2c26-4cf9-a0e0-56950b638135)



A restart of VSCode window can help resolve that issue, but it is considered "too heavy".

**Comment by gcampbell-msft** [comment, marked as answer]
@zchrissirhcz What about your scenario requires the setup you have, which adds complexity and multiple tasks, rather than simply relying on "CMake: Configure" and the typical flow from the extension. It seems to me from your task set that you're not actually "Configuring" with our extension which is actually what integrates with the Intellisense extension (ms-vscode.cpptools).

**Reply by zchrissirhcz** [reply]
I think passing customized CMake options/cache variables (say, `-DHELLO_TYPE=HAPPY` )during the cmake configure process is a common scenario. However, when using the CMake Tools extension, I habitually click the 'Build' button in the bottom status bar of VSCode with my mouse, which does not pass the customized CMake options.

I know that CMake Tools extension provide a GUI interface to modify cache variables. But in that case, I have to invoke CMake configure for a second run. **I hope I can run CMake configure process only once, with mouse clicking, with customized cmake options/cache variables specified**. What I know is to use `tasks.json`.

**Reply by gcampbell-msft** [reply]
This would be an excellent use case for CMakePresets! This allows you to construct and control the environment and cache variables for each configure preset. Then, whenever you configure, this environment can be easily controlled. 

https://cmake.org/cmake/help/latest/manual/cmake-presets.7.html

**Reply by zchrissirhcz** [reply]
After writing several minimal examples of cmake-presets, I figured out CMakePresets.json is the right one for cache variables.
